detectors/node/resource-detector-service/src/service.ts

```diff
@@ -0,0 +1,45 @@
+import { ResourceAttributes as ResourceAttributesKeys } from '@opentelemetry/semantic-conventions';
+import { Resource, defaultServiceName, ResourceAttributes } from '@opentelemetry/resources';
+import { SyncDetector, SyncDetectorToDetector } from 'opentelemetry-resource-detector-sync-api';
+import { v4 as uuidv4 } from 'uuid';
+import * as fs from 'fs';
+
+// set as global to make sure it's the same on any invocation even for multiple
+// instances of ServiceSyncDetector
+const instanceId = uuidv4();
+
+class ServiceSyncDetector implements SyncDetector {
+    detect(): Resource {
+        const packageJson = this.loadJsonFile('package.json');
+        const attributes: ResourceAttributes = {
+            [ResourceAttributesKeys.SERVICE_INSTANCE_ID]: instanceId,
+            [ResourceAttributesKeys.SERVICE_NAME]: this.getServiceName(packageJson),
+        };
+        const serviceVersion = packageJson?.version;
+        if (serviceVersion) {
+            attributes[ResourceAttributesKeys.SERVICE_VERSION] = serviceVersion;
+        }
+        return new Resource(attributes);
+    }
+
+    getServiceName(packageJson: any): string {
+        const fromEnv = process.env.OTEL_SERVICE_NAME;
+        if (fromEnv) return fromEnv;
+
+        const fromPackageJson = packageJson?.name;
+        if (fromPackageJson) return fromPackageJson;
+
+        return defaultServiceName();
+    }
+
+    loadJsonFile(path: string): any {
+        try {
+            return JSON.parse(fs.readFileSync(path).toString());
```

Comment:
I don't know why, but it gives "Error: Cannot find module 'package.json'" when I try to `require('package.json')`.
Probably the `require` base path is different somehow vs the `fs.readFileSync` alternative (which I believe can also bump into some edge cases where the process is not started from the service root)
If you have experience on how to write it so it works, that would be great :)

Reply:
That's because it is looking for a package named `package.json`. `./package.json` tells `require` to look for the file.

Both solutions are prone to errors from running the process from the outside of the project root folder.
I don't think there's a way to get that without letting the user specify the root OR the path to the main `package.json`.

Best way I could come up with to more or less consistently find what you are looking for is this:

```js
const path = require('path');

const findRootPackageJson = () => {
	const paths = require.main.paths.map((p) => {
		return path.resolve(p, '..', 'package.json');
	});
	for (const path of paths) {
		try {
			return {
				path,
				contents: require(path),
			};
		} catch {}
	}

	return {};
}

console.log(require.main);
console.log(findRootPackageJson());
```